### PR TITLE
feat(formatter): aceitar CNPJ alfanumerico (Reforma Tributaria 2026-07-01)

### PR DESCRIPTION
**Pull request by rhfranzoni**
## Resumo

A SEFAZ vai aceitar **CNPJ alfanumerico** a partir de **2026-07-01** (LC 214/2025, IN RFB 2.229/2024). 12 primeiros caracteres podem conter letras maiusculas; 2 ultimos seguem como digitos verificadores numericos.

Hoje o `Formatador.FormatarCnpj`/`FormatarCpfCnpj` usa regex `\d` que rejeita letras — efeito pratico: a partir de 2026-07-01 o DANFE/cupom passa a renderizar CNPJ alfanumerico **sem mascara visual**, prejudicando legibilidade fiscal.

### `DanfeSharp/Formatador.cs`
- **Regex CNPJ**: 4 primeiros grupos passam de `\d` para `[0-9A-Z]`. Ultimo grupo (DV) permanece `\d`.
- **`FormatarCpfCnpj`**: reescrita para decidir por **comprimento da string limpa** antes da regex (elimina ambiguidade entre CPF e CNPJ alfanumerico).
- **CPF regex**: inalterada (CPF segue numerico por norma RFB).

### `DanfeSharp.Test/FormatadorTests.cs` (novo)
30+ casos via `DataTestMethod` / `DataRow`:
- Regressao: CNPJs/CPFs numericos legados.
- Novos casos: CNPJ alfanumerico em multiplas posicoes (`12ABC345000188`, `98XYZ123000158`).
- Rejeicao: lowercase, DV alfabetico, comprimento invalido.
- Edge cases: null, vazio, espacos, ja-formatado.

## Validacao

- DLL principal (`DanfeSharp.csproj`) compila sem erros.
- Validacao via `dotnet-script` chamando `Formatador.FormatarCnpj` / `FormatarCpfCnpj` com 13 vetores: **13/13 OK**.

### Ressalva
O test project `DanfeSharp.Test` ja apresentava erros preexistentes de compilacao em `FabricaFake.cs` e `DanfeTest.cs` (`DanfeCCC` nao encontrado, conversao double->decimal) na branch main antes deste PR. Esses erros nao foram introduzidos aqui e nao tocam no codigo modificado. Como follow-up separado: corrigir o test project para que `dotnet test` rode completo.

## Issues vinculadas

- `Closes #35` apos serem feitos os outros PRs derivados (esta e a primeira etapa).
- Relacionada `#33` (barcode CODE-128A/C com CNPJ alfanumerico) — esta PR resolve o requisito de tipagem; barcode em si segue como track paralelo.

## Test plan

- [x] DLL compila
- [x] dotnet-script valida 13/13 vetores (CNPJ numerico, alfanumerico, CPF, edge cases)
- [ ] CI do fork (se configurado) executa o test project corrigido
- [ ] Revisor humano valida a regex e o switch por length

## Breaking changes

Nenhuma:
- CNPJs numericos legados formatam exatamente como antes.
- Comprimento invalido devolve input trimado (sem exception) — mesmo contrato historico.
- Null / vazio retornam string vazia — mesmo contrato historico.

Pessoas consumindo `FormatarCnpj` precisam estar preparadas para receber **strings com letras** no retorno a partir de 2026-07-01 — mas o tipo de retorno (`string`) e o formato visual (`XX.XXX.XXX/XXXX-XX`) sao os mesmos.

Generated with Claude Code (https://claude.com/claude-code)

---
Relates to #33 (requisito de tipagem do CNPJ alfanumérico; barcode CODE-128A/C segue como track paralelo — issue permanece aberta).

### CODE REVIEW

File: DanfeSharp/Formatador.cs [hunk 1]
```diff
@@ -122,33 +125,34 @@ public static String FormatarCpf(String cpf)
...
+            String trimmed = cpfCnpj.Trim();
+            String clean = trimmed.Replace(".", String.Empty).Replace("-", String.Empty).Replace("/", String.Empty);
+
+            switch (clean.Length)
             {
-                result = String.Empty;
+                case 11:
+                    return FormatarCpf(clean);
+                case 14:
+                    return FormatarCnpj(clean);
```
[the remark this answers is not in the document]
Comment (rhfranzoni) [reply DanfeSharp/Formatador.cs]: Bom ponto — enderecado em 7a41909. Agora o dispatch valida `clean` contra a regex correspondente ANTES de chamar `FormatarCpf`/`FormatarCnpj`. Conteudo invalido (lowercase em CNPJ alfanumerico, DV alfabetico, etc.) cai no fallback `return trimmed`, preservando a pontuacao visivel do input original.

Validacao manual via `dotnet-script` contra a DLL recem-buildada cobriu 13 cenarios — todos OK incluindo o caso exato do comentario (`"12.abc.345/0001-88"` -> retorna identico).

Novos testes em `FormatadorTests.cs`:
- `FormatarCpfCnpj_InvalidoComMascara_PreservaPontuacaoDoInput` (4 cenarios)
- `FormatarCpfCnpj_InvalidoSemMascara_DevolveInputInalterado` (2 cenarios)

File: DanfeSharp.Test/DanfeSharp.Test.csproj [hunk 2]
```diff
@@ -61,6 +61,7 @@
     <Compile Include="DanfeXmlTests.cs" />
     <Compile Include="Extentions.cs" />
     <Compile Include="FabricaFake.cs" />
+    <Compile Include="FormatadorTests.cs" />
```
Comment (john182) [inline DanfeSharp.Test/DanfeSharp.Test.csproj:65]: ⚠️ **Importante — estes testes não rodam via `dotnet test`.** O assembly `DanfeSharp.Test` não compila: `DanfeTest.cs:63` referencia `DanfeCCC`, classe que não existe no repo (confirmei). Enquanto o test project não buildar, os 36 casos deste arquivo não executam em CI nem localmente — só foram validados por `dotnet-script`, que exercita o `Formatador` mas **não** o arquivo de teste em si.

💡 Recomendo corrigir o test project **neste PR** (ou em PR pré-requisito mergeado antes deste). Caso contrário, são testes mortos até o follow-up — e o risco é que nunca rodem e a regex sofra regressão silenciosa.

File: DanfeSharp/Formatador.cs [hunk 3]
```diff
@@ -29,7 +29,10 @@ static Formatador()
         public const String FormatoNumeroNF = @"000\.000\.000";
 
         public const String CEP = @"^(\d{5})\-?(\d{3})$";
-        public const String CNPJ = @"^(\d{2})\.?(\d{3})\.?(\d{3})\/?(\d{4})\-?(\d{2})$";
+        // CNPJ aceita letras maiúsculas nos 12 primeiros caracteres a partir de 2026-07-01
+        // (Reforma Tributária, LC 214/2025 e IN RFB 2.229/2024). Os 2 últimos seguem como
+        // dígitos verificadores numéricos. CPF permanece estritamente numérico.
+        public const String CNPJ = @"^([0-9A-Z]{2})\.?([0-9A-Z]{3})\.?([0-9A-Z]{3})\/?([0-9A-Z]{4})\-?(\d{2})$";
```
Comment (john182) [inline DanfeSharp/Formatador.cs:35]: ✅ Regex correta conforme a norma: 12 primeiros `[0-9A-Z]`, DV final `\d{2}` numérico.

💡 Decisão de design a confirmar: lowercase é **rejeitado** (volta sem máscara) em vez de normalizado para maiúsculo. É defensável — a forma canônica SEFAZ é maiúscula — e os testes assertam isso explicitamente. Só vale garantir que nenhum caminho upstream entregue o CNPJ em minúsculas ao formatador; se entregar, o DANFE renderiza sem máscara silenciosamente. Se houver essa dúvida, um `.ToUpperInvariant()` no `clean` resolveria de forma conservadora.

File: DanfeSharp/Formatador.cs [hunk 4]
```diff
@@ -122,33 +125,47 @@ public static String FormatarCpf(String cpf)
...
+            String trimmed = cpfCnpj.Trim();
+            String clean = trimmed.Replace(".", String.Empty).Replace("-", String.Empty).Replace("/", String.Empty);
+
+            // Valida o conteúdo limpo contra a regex correspondente ANTES do dispatch.
+            // Sem essa validação, entradas como "12.abc.345/0001-88" (lowercase rejeitado
+            // pelo formato CNPJ alfanumérico) cairiam no dispatch por comprimento, o
+            // FormatarCnpj retornaria a string crua sem máscara, e a pontuação visível
+            // do input original seria perdida. Devolver `trimmed` para input inválido
+            // alinha com o contrato histórico: regex miss => input inalterado.
+            if (clean.Length == 11 && Regex.IsMatch(clean, CPF))
```
Comment (john182) [inline DanfeSharp/Formatador.cs:156]: Nit (não bloqueia): `clean.Length == 11` é redundante com a regex `CPF`, que é ancorada (`^...$`) e já exige exatamente 11 caracteres sobre a string limpa (idem `== 14` para CNPJ). Mantê-lo como fast-path/intenção explícita é aceitável — só registrando que não é necessário para corretude.

**Review by john182 — COMMENTED** [review]
## 🔍 Review — PR #36 `feat(formatter): CNPJ alfanumérico`

Mudança cirúrgica e correta. Validei o `Formatador.cs` completo no head do PR (incl. `InternalRegexReplace`, `FormatarCnpj/Cpf`) — a lógica de produção está sólida e **retrocompatível de fato**. Comentários inline nos pontos específicos.

### ✅ Pontos fortes
- Regex alinhada à norma: 12 primeiros `[0-9A-Z]` + DV `\d{2}` numérico (LC 214/2025).
- **Retrocompatível**: CNPJ/CPF numéricos legados formatam idênticos; input inválido volta inalterado — mesmo contrato do `InternalRegexReplace`, que em regex-miss retorna o input.
- O refactor (dispatch por comprimento → **depois** regex) preserva o contrato histórico "regex miss ⇒ input intocado" e ainda ganhou teste de regressão dedicado (`...PreservaPontuacaoDoInput`).
- Cobertura de edge cases excelente (36 casos): null, vazio, espaços, lowercase, DV alfabético, com/sem máscara, idempotência.
- Disclosure honesto do test project quebrado + análise de breaking changes.
- `Compile Include` adicionado corretamente (csproj old-style, não SDK).

### ⚠️ Importante
Os 36 testes **não executam** — o assembly `DanfeSharp.Test` não compila (`DanfeCCC` inexistente em `DanfeTest.cs:63`, confirmado). Detalhe inline na csproj. É o único ponto que eu trataria antes do merge: teste que não roda não protege contra regressão.

### 💡 Menores
- Nomes dos testes seguem `Metodo_Cenario_Esperado` (MSTest) em vez de Given/When/Then — **ok**, segue a convenção do repo.
- Redundância na checagem de comprimento (inline, não bloqueia).

**Veredito:** lógica de produção aprovada. Recomendo destravar a execução do test project para que a rede de proteção realmente exista.

File: DanfeSharp.Test/DanfeSharp.Test.csproj (hunk 2 above)
Comment (rhfranzoni) [reply DanfeSharp.Test/DanfeSharp.Test.csproj:65]: Corrigido neste PR (commits c77b7ca + bdd18d4). O assembly `DanfeSharp.Test` agora **compila** (gera `DanfeSharp.Test.dll`):

1. `DanfeTest.cs`: `DanfeCCC` (inexistente) → **`DanfeCartaCorrecao(DanfeEventoViewModel)`** — a classe correta de CC-e, cujo construtor aceita o `DanfeEventoViewModel` retornado por `DanfeEventoViewModelCreator.CriarDeArquivoXml`.
2. Ao mesclar a `main`, surgiu um **2º bloqueio pré-existente**: `DanfeSimplificadoTipo2Tests.cs` usava `Extentions` não-qualificado, que resolvia para `DanfeSharp.Test.Extentions` (sem `UrlNFeConsulta`) em vez de `DanfeSharp.Extentions` (internal, via `InternalsVisibleTo`) → `CS0117`. Qualifiquei para `DanfeSharp.Extentions.UrlNFeConsulta`.
3. Conflito no `DanfeSharp.Test.csproj` resolvido pela **união** dos `<Compile Include>` (FormatadorTests deste PR + os 3 arquivos de teste vindos da main).

Obs.: o adapter **MSTest 1.1.11 (legacy, net461)** não executa via `dotnet test` no meu ambiente local — a execução dos casos fica para o runner do CI, mas o blocker que você apontou (não buildar) está resolvido.

File: DanfeSharp/Formatador.cs (hunk 3 above)
Comment (rhfranzoni) [reply DanfeSharp/Formatador.cs:35]: Obrigado. Sobre o ponto de design (lowercase): mantida a **rejeição** (retorna sem máscara), de propósito — a forma canônica do CNPJ alfanumérico na SEFAZ é **maiúscula** (base-36, letras A–Z), e os testes assertam isso explicitamente. O upstream trabalha o CNPJ nessa forma canônica, então não há caminho previsto entregando minúsculo ao formatador. Se surgir tal caminho, adotamos o `.ToUpperInvariant()` conservador que você sugeriu (normalizar em vez de rejeitar). Registrado.

File: DanfeSharp/Formatador.cs (hunk 4 above)
Comment (rhfranzoni) [reply DanfeSharp/Formatador.cs:156]: Obrigado pelo registro. Mantido o `clean.Length == 11`/`== 14` como **fast-path/intenção explícita** (como você marcou, é aceitável): não é necessário para corretude dado o ancoramento `^...$` da regex, mas documenta a intenção e evita invocar a regex desnecessariamente para tamanhos fora do esperado.
